[RISCV] Add 16 bit GPR sub-register for Zhinx.

topperc opened the pull request:
This patches adds a 16 bit register class for use with Zhinx instructions. This makes them more similar to Zfh instructions and allows us to only spill 16 bits.

I've added CodeGenOnly instructions for load/store using GPRF16 as that gave better results than insert_subreg/extract_subreg. I'm using FSGNJ for GPRF16 copy with Zhinx as that gave better results. Zhinxmin will use ADDI+subreg operations.

Function arguments use this new GPRF16 register class for f16 arguments with Zhinxmin. Eliminating the need to use RISCVISD::FMV* nodes.

I plan to extend this idea to Zfinx next.

After that, I want to try to extend this to 32 bit integer W instructions. My thought is that we can arrange to have all writes to the 32 bit GPR guarantee sign extension similar to how Mips64 is handled. Unfortunately, we are missing some W instructions in Zba and Zbs that would make this straightforward. I want to see if this is more viable than my previous attempt at making i32 a legal type which kept it inside the existing 64-bit GPR class.

topperc committed:
[RISCV] Add 16 bit GPR sub-register for Zhinx.

This patches adds a 16 bit register class for use with Zhinx instructions.
This makes them more similar to Zfh instructions and allows us to
only spill 16 bits.

I've added CodeGenOnly instructions for load/store using GPRF16 as that
gave better results than insert_subreg/extract_subreg. I'm using FSGNJ for
GPRF16 copy with Zhinx as that gave better results. Zhinxmin will use
ADDI+subreg operations.

Function arguments use this new GPRF16 register class for f16 arguments
with Zhinxmin. Eliminating the need to use RISCVISD::FMV* nodes.

I plan to extend this idea to Zfinx next.

After that, I want to try to extend this to 32 bit integer W instructions.
My thought is that we can arrange to have all writes to the 32 bit
GPR guarantee sign extension similar to how Mip64 is handled. Unfortunately,
we are missing some W instructions in Zba and Zbs that would make this
straightforward.

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVRegisterInfo.td:
```diff
@@ -83,41 +83,81 @@ def sub_gpr_odd  : SubRegIndex<32, 32> {
+  def X31_H : RISCVReg<31,"x31", ["t6"]>;
+  }
+
+  let SubRegIndices = [sub_16] in {
```
Do we need `CoveredBySubRegs`?

topperc replied:
It's not covered by subregs. There's no sub reg for the upper bits.

preames commented:
This goes beyond what I can meaningfully review, but I'll note that I see nothing to particularly object to here.  

Naming wise, _H is, I assume, for half.  Since we'll probably use these for bf16, and possible someday i16, maybe pick a different name suffix?

jrtc27 commented:
FLH and LH both use half as the term, and the format of the float shouldn’t affect the load or store, just the interpretation for arithmetic purposes, so _H seems like the appropriate thing to use for me?

preames commented:
> FLH and LH both use half as the term, and the format of the float shouldn’t affect the load or store, just the interpretation for arithmetic purposes, so _H seems like the appropriate thing to use for me?

Good point, naming suggestion withdrawn.

topperc committed:
Merge remote-tracking branch 'origin/main' into pr/zhinx-subreg

topperc committed:
Merge remote-tracking branch 'origin/main' into pr/zhinx-subreg

topperc committed:
fixup! Remove unneeded DecoderMethod =

asb reviewed (approved):
LGTM in that the reasoning makes sense to me and I can't see problems from going through the code. That said, it's been a while since I added a new register class.

I think it would be nice to have a brief comment in RISCVRegisterInfo.td explaining why we're defining this register class and how it's used. Perhaps that makes more sense once you make similar changes for zfinx, but in general I find such in-tree documentation of intent really helpful.

You say that fsgnj gives better results vs mv - I'm assuming this is for real world code? Because for the modified in-tree tests, surely using `mv` where possible is better due to being potentially compressible? Happy to go with your judgement, but it probably deserves a TODO/FIXME somewhere like the tests to note that a semantically equivalent mv would be more compressible (and so someone sufficiently motivated might later pick it up).

EDIT to add:
> After that, I want to try to extend this to 32 bit integer W instructions. My thought is that we can arrange to have all writes to the 32 bit GPR guarantee sign extension similar to how Mips64 is handled. Unfortunately, we are missing some W instructions in Zba and Zbs that would make this straightforward. I want to see if this is more viable than my previous attempt at making i32 a legal type which kept it inside the existing 64-bit GPR class.

This definitely sounds like an interesting direction. And if we can demonstrate adding those missing instructions makes life easier for the compiler, it's probably worth at least discussing in e.g. the scalar efficiency working group.

topperc commented:
> You say that fsgnj gives better results vs mv - I'm assuming this is for real world code? Because for the modified in-tree tests, surely using `mv` where possible is better due to being potentially compressible? Happy to go with your judgement, but it probably deserves a TODO/FIXME somewhere like the tests to note that a semantically equivalent mv would be more compressible (and so someone sufficiently motivated might later pick it up).

The subreg operations needed to use the existing ADDI seem to pessimize Machine Copy Propagation on some of the lit tests. I think this is because some implicit operands get added and MachineCopyPropagation doesn't know how those implicit operands relate to the explicit operands so it conservatively gives up. I have not done any testing of this on real world code. I suspect a GPRF16 CodeGenOnly copy instruction may fix this issue, but haven't checked yet.

topperc committed:
Merge remote-tracking branch 'origin/main' into pr/zhinx-subreg

topperc committed:
fixup! Add comment

topperc committed:
fixup! Add Zhinx load/store to RISCVMakeCompressible

topperc committed:
fixup! Add compressed load/store CodeGenOnly instructions.

topperc committed:
Merge remote-tracking branch 'origin/main' into pr/zhinx-subreg

topperc committed:
Merge remote-tracking branch 'origin/main' into pr/zhinx-subreg

topperc committed:
fixup! Add a pseudo MV instruction for copy so we can compress it.

topperc committed:
fixup! Add load/store to other locations that reference RISCV::LH/SH

topperc committed:
fixup! More updates for RISCVMakeCompressible.

topperc committed:
fixup! testing and fixes for RISCVMakeCompressible

topperc committed:
Merge remote-tracking branch 'origin/main' into pr/zhinx-subreg

topperc committed:
fixup! clang-format

topperc commented:
@asb can you re-review this? I added compressed versions of the load/store CodeGenOnly instructions, a pseudo for mv, updated MakeCompressible, MergeBaseOffset.

asb commented on llvm/lib/Target/RISCV/RISCVCallingConv.cpp:
```diff
@@ -139,6 +139,23 @@ ArrayRef<MCPhysReg> RISCV::getArgGPRs(const RISCVABI::ABI ABI) {
+  static const MCPhysReg ArgIGPRs[] = {RISCV::X10_H, RISCV::X11_H, RISCV::X12_H,
+                                       RISCV::X13_H, RISCV::X14_H, RISCV::X15_H,
+                                       RISCV::X16_H, RISCV::X17_H};
+  // The GPRs used for passing arguments in the ILP32E/ILP64E ABI.
```
Typo: ILP64E => LP64E (I've just committed a fix to other similar instances in this file).

asb commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -1505,6 +1520,9 @@ unsigned RISCVInstrInfo::getInstSizeInBytes(const MachineInstr &MI) const {
 
   switch (Opcode) {
+  case RISCV::PseudoMV_FPR16INX:
+    // MV is always compressible.
```
This comment made me double check, so sharing that process: Although we'd expect `c.li` to be used instead, `c.mv` can't handle an X0 operand. But from what I can see, we expand to ADDI and we have the obvious CompressPat from ADDI with X0 operand to c.li.

Perhaps expand the comment to "MV is always compressible (to either c.mv or c.li rd, 0)."

asb reviewed (approved):
LGTM! Sorry for missing your request to re-review. I'm still missing the setup I had with Phabricator Herald :( But I'll try to improve my workflow.

topperc committed:
Merge remote-tracking branch 'origin/main' into pr/zhinx-subreg

topperc committed:
fixup! Address comments.
